Added HAPI like DEFAULT tenant support, added grant access on all ten…

mario-degenhardt-snkeos opened the pull request:
Issue #, if available:

Description of changes:
setTenantId.ts:
added "DEFAULT" tenant, Machine2Machine authenticationbased grantAllTenants handling
merged custom claim handling to be compatible with our cognito:groups
Added additional multi tenancy tests
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

mario-degenhardt-snkeos committed:
Added HAPI like DEFAULT tenant support, added grant access on all tenants by specific scope (required for machine 2 machine auth flow use cases)

Ensure correct urls including tenant id if enabled in transaction and search set bundle responses

organom commented on src/app.multitenancy.test.ts:
```diff
@@ -0,0 +1,419 @@
+const wrongTenantId: string = '1234567890';
+const anyTenantId: string = '09876543';
+
+const request = require('supertest');
```
should be an import as the rest

mario-degenhardt-snkeos replied:
The AWS guys also use require('supertest') in their tests, see app.test.ts

organom replied:
We don't need to copy something if its wrong :)  if its a global require (and all the others are import), then it should be converted into an import (in webstorm when selecting the line, right click -> show context actions -> convert require to import). 

Don't forget to run the tests just to make sure

mario-degenhardt-snkeos replied:
Then the following dependencies needs to be added:
     "@types/supertest": "^2.0.12",
    "@types/superagent": "4.1.13",  <= Workaround: Here a fixed version needs to set, because  the 4.1.14 has an export buffer form Blob bug
and "@types/node": "^12 needs to be updated to "@types/node": "^14.0.1",

I guess the AWS guys did that way to avoid these dependencies.

organom replied:
Hummm, I see, then typescript also tries to import the types with the import command. ok, then lets leave it as they did :)

organom commented on src/router/__mocks__/authorizationService.ts:
```diff
@@ -0,0 +1,61 @@
+}
+const AuthorizationService: Authorization = class {
+    static async verifyAccessToken(request: VerifyAccessTokenRequest): Promise<KeyValueMap> {
+        if (!expectedtokenDecoded) {
```
would look better like:

```
return expectedtokenDecoded ? expectedtokenDecoded : practitionerDecoded;
```

or if you want to keep the if, then inverted:
```
if(expectedtokenDecoded) {
   return expectedtokenDecoded;
}
return {
    sub: 'fake',
    'cognito:groups': ['practitioner'],
    name: 'not real',
    iat: 1516239022,
};
```

organom commented on src/router/middlewares/setTenantId.ts:
```diff
@@ -48,6 +48,92 @@ const getTenantIdFromAudClaim = (audClaim: any, baseUrl: string): string | undef
+ * @param fhirConfig The config, which includes multi-tenant setup.
+ * @returns true access granted, false denied
+ */
+function grantAccessForAllTenants(userIdentity: any, fhirConfig: FhirConfig): boolean {
```
wrong naming, it does not grant access.  Alternatives: `hasAccessToAllTenants` or `evaluateAccessForAllTenants`

organom commented on src/router/middlewares/setTenantId.ts:
```diff
@@ -48,6 +48,92 @@ const getTenantIdFromAudClaim = (audClaim: any, baseUrl: string): string | undef
+ * @param tenantIdFromPath The tenant id url parameter to test
+ * @returns true access granted, false denied
+ */
+function grantAccessForCustomClaim(
```
naming, like above. in this case it returns a string not a bool, so like `getCustomClaim` or `getTenantIdFromCustomClaim`

organom commented on src/router/middlewares/setTenantId.ts:
```diff
@@ -48,6 +48,92 @@ const getTenantIdFromAudClaim = (audClaim: any, baseUrl: string): string | undef
+ * @returns true access granted, false denied
+ */
+function grantAccessForAllTenants(userIdentity: any, fhirConfig: FhirConfig): boolean {
+    if (fhirConfig.multiTenancyConfig?.grantAccessAllTenantsScope) {
```
`return fhirConfig.multiTenancyConfig?.grantAccessAllTenantsScope && userIdentity.scope?.includes(fhirConfig.multiTenancyConfig?.grantAccessAllTenantsScope)`

organom commented on src/router/middlewares/setTenantId.ts:
```diff
@@ -48,6 +48,92 @@ const getTenantIdFromAudClaim = (audClaim: any, baseUrl: string): string | undef
+ * @param tenantIdFromCustomClaim Possible tenant id values from custom claim
+ * @param prefix The optional prefix for tenantIdClaimValue
+ * @param tenantIdFromPath The tenant id url parameter to test
+ * @returns true access granted, false denied
```
wrong

organom commented on src/router/middlewares/setTenantId.ts:
```diff
@@ -48,6 +48,92 @@ const getTenantIdFromAudClaim = (audClaim: any, baseUrl: string): string | undef
+ * @param tenantId The tenant id url parameter to test
+ * @returns true access granted, false denied
+ */
+function grantAccessForAudClaim(tenantIdFromAudClaim: any | undefined, tenantIdFromPath: string): string | undefined {
```
wrong naming

organom commented on src/router/middlewares/setTenantId.ts:
```diff
@@ -48,6 +48,92 @@ const getTenantIdFromAudClaim = (audClaim: any, baseUrl: string): string | undef
+ * Evaluates if tenant id url param value matches with aud claim value, or not.
+ * @param tenantIdFromAudClaim Possible tenant id value from aud claim
+ * @param tenantId The tenant id url parameter to test
+ * @returns true access granted, false denied
```
wrong return defined

organom commented on src/router/middlewares/setTenantId.ts:
```diff
@@ -48,6 +48,92 @@ const getTenantIdFromAudClaim = (audClaim: any, baseUrl: string): string | undef
+/**
+ * Evaluates if tenant id url param value matches with aud claim value, or not.
+ * @param tenantIdFromAudClaim Possible tenant id value from aud claim
+ * @param tenantId The tenant id url parameter to test
```
wrong parameter name

organom commented on src/router/middlewares/setTenantId.ts:
```diff
@@ -48,6 +48,92 @@ const getTenantIdFromAudClaim = (audClaim: any, baseUrl: string): string | undef
+    }
+
+    // Multiple possible tenant id values in custom claim is only supported, if a tenantId from url path is available
+    if (tenantIdFromCustomClaimAsArray.length > 1 && tenantIdFromPath !== undefined) {
```
`if (tenantIdFromCustomClaimAsArray.length > 1 && tenantIdFromPath)`

organom commented on src/router/middlewares/setTenantId.ts:
```diff
@@ -48,6 +48,92 @@ const getTenantIdFromAudClaim = (audClaim: any, baseUrl: string): string | undef
+        const tenantIdCandidate: string = tenantIdFromCustomClaimAsArray[0]!;
+        if (
+            tenantIdRegex.test(tenantIdCandidate) &&
+            (tenantIdFromPath === undefined || tenantIdFromPath === tenantIdCandidate)
```
`(!tenantIdFromPath || tenantIdFromPath === tenantIdCandidate)`

organom commented on src/router/middlewares/setTenantId.ts:
```diff
@@ -48,6 +48,92 @@ const getTenantIdFromAudClaim = (audClaim: any, baseUrl: string): string | undef
+        }
+    }
+
+    return undefined;
```
no need for the above

mario-degenhardt-snkeos replied:
If I remove it, I get the ESlint error consistent-return, which don't want to disable.

organom commented on src/router/middlewares/setTenantId.ts:
```diff
@@ -48,6 +48,92 @@ const getTenantIdFromAudClaim = (audClaim: any, baseUrl: string): string | undef
+    if (tenantIdClaimValue.startsWith(prefix)) {
+        return tenantIdClaimValue.substring(prefix.length);
+    }
+    return undefined;
```
no need for the return undefined. By default it'll be undefined

mario-degenhardt-snkeos replied:
If I remove it, I get the ESlint error consistent-return, which don't want to disable.

organom commented on src/router/middlewares/setTenantId.ts:
```diff
@@ -48,6 +48,92 @@ const getTenantIdFromAudClaim = (audClaim: any, baseUrl: string): string | undef
+    ) {
+        return tenantIdFromAudClaim;
+    }
+    return undefined;
```
no need for above

mario-degenhardt-snkeos replied:
If I remove it, I get the ESlint error consistent-return, which don't want to disable.

mario-degenhardt-snkeos committed:
Fixed review remarks

- renamed grantAccessForAudClaim to filterTenantId
- renamed grantAccessForCustomClaim to getTenantIdFromCustomClaim
- renamed grantAccessForAllTenants to evaluateAccessForAllTenants
- reused filterTenantId in getTenantIdFromCustomClaim
- fixed documentation
- adapted to suggested coding style

organom commented:
Outside of the inline comment, it looks good. Please fix the import and the rest is good :)
